aten/src/ATen/templates/ATenDispatch.h

```diff
@@ -0,0 +1,89 @@
+#pragma once
+
+#include <c10/core/Backend.h>
+#include <unordered_map>
+
+// This dispatch class serves as a replacement for Type based dispatch.
+// It stores a function for each native function and backend pair.
+// Functions that are the same for all backends can be registered under
+// Backend::Undefined.
+
+namespace at {
+
+template<class FuncType>
+class CAFFE2_API ATenOp {};
+template<class Return, class... Params>
+class CAFFE2_API ATenOp<Return (Params...)> {
+  using FnPtr = Return (*)(Params...);
+  using WrapperPtr = Return (*)(FnPtr, Params...);
+ public:
+  Return operator()(Params... params) const {
+    if (wrapper_) {
+      return (*wrapper_)(fn_, params...);
+    }
+    return (*fn_)(params...);
+  }
+
+ private:
+  explicit ATenOp(void* fn)
+  : fn_(reinterpret_cast<FnPtr>(fn)), wrapper_(nullptr) {}
+  explicit ATenOp(void* fn, void* wrapper)
+  : fn_(reinterpret_cast<FnPtr>(fn)), wrapper_(reinterpret_cast<WrapperPtr>(wrapper)) {}
+  friend class ATenDispatch;
+
+  FnPtr fn_;
+  WrapperPtr wrapper_;
+};
+
+class CAFFE2_API ATenDispatch {
+ public:
+  template<class FuncType>
+  ATenDispatch& registerOp(Backend backend, const char* schema, FuncType* fn) {
+   auto id = getSchemaId(schema);
+   function_table[static_cast<int64_t>(backend)][id] = reinterpret_cast<void*>(fn);
```

Comment:
You don't need a `reinterpret_cast` to cast a pointer into a void pointer. (It's only needed the other way.)

Reply:
So... no error checking, I guess?